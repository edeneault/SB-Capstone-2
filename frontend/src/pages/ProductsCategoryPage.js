import React, { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { Row, Col, Container } from "react-bootstrap";
import Product from "../components/Product";
import Loader from "../components/Loader";
import Paginate from "../components/Paginate";
import CategoryCarousel from "../components/CategoryCarousel";
import { listCategoryProducts } from "../actions/productActions";

import { toast } from "react-toastify";
import "react-toastify/dist/ReactToastify.css";

toast.configure();

const ProductsCategoryPage = ({ match }) => {
  const category = match.params.category;
  const pageNumber = match.params.pageNumber || 1;
  const dispatch = useDispatch();

  const productCategory = useSelector((state) => state.productCategory);
  const { loading, error, products, page, pages } = productCategory;

  useEffect(() => {
    dispatch(listCategoryProducts(category, pageNumber));
  }, [dispatch, category, pageNumber]);

  window.scroll({
    top: 0,
    left: 0,
    behavior: "smooth",
  });

  return (
    <>
      <CategoryCarousel category={category} />
      <Container fluid className='px-5 pt-3'>
        {loading ? (
          <Loader />
        ) : error ? (
          toast(`${error}`, { type: "error" })
        ) : (
          <>
            <Row>
              {products.map((product) => (
                <Col sm={12} md={6} lg={4} xl={3} key={product._id}>
                  <Product product={product} />
                </Col>
              ))}
            </Row>
            <Paginate
              pages={pages}
              page={page}
              category={category ? category : ""}
            />
          </>
        )}
      </Container>
    </>
  );
};

export default ProductsCategoryPage;
