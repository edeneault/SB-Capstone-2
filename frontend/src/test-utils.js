import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";

// create a customRender that wraps the UI in a memory Router

const MemoryRouterWithInitialRoutes = ({ children, initialRoutes }) => {
  return <MemoryRouter initialEntries={initialRoutes}>{children}</MemoryRouter>;
};

const customRender = (ui, options) => {
  const initialRoutes =
    options && options.initialRoutes ? options.initialRoutes : ["/"];

  return render(ui, {
    wrapper: (args) =>
      MemoryRouterWithInitialRoutes({
        ...args,
        initialRoutes,
      }),
    ...options,
  });
};

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
