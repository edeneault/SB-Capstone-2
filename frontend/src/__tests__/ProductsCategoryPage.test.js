import React from "react";
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";
import thunk from "redux-thunk";
import { render } from "../test-utils";
import { initialState } from "../store-mock";
import ProductsCategoryPage from "../pages/ProductsCategoryPage.js";

const mockStore = configureMockStore([thunk]);

describe("test", () => {
  let category;
  beforeEach(() => {
    jest.spyOn(global.Math, "random").mockReturnValue(0.123456789);
    category = "espresso";
  });

  afterEach(() => {
    jest.spyOn(global.Math, "random").mockRestore();
  });
  test("renders ProductsCategoryPage without crashing", async () => {
    const store = mockStore(initialState);
    const match = {
      params: {
        category: category,
      },
    };

    const { getByText } = render(
      <Provider store={store}>
        <ProductsCategoryPage match={match} />
      </Provider>,
      { initialRoutes: [`/allproducts`] },
    );

    expect(getByText(/~ category espresso ~/i)).toBeInTheDocument();
  });

  test("matches snapshot", function () {
    const store = mockStore(initialState);
    const match = {
      params: {
        category: category,
      },
    };

    const { asFragment } = render(
      <Provider store={store}>
        <ProductsCategoryPage match={match} />
      </Provider>,
      { initialRoutes: [`/allproducts`] },
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
