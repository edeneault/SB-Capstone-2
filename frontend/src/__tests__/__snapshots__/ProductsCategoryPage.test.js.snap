// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test matches snapshot 1`] = `
<DocumentFragment>
  <h2
    class="carousel-title m-0 p-3 text-dark fs-1"
  >
    ~ Category espresso ~
  </h2>
  <div
    class="react-multi-carousel-list  fade-in"
  >
    <ul
      class="react-multi-carousel-track "
      style="transition: none; overflow: unset; transform: translate3d(0px,0,0);"
    />
  </div>
  <div
    class="px-5 pt-3 container-fluid"
  >
    <div
      class="row"
    >
      <div
        class="col-xl-3 col-lg-4 col-md-6 col-sm-12"
      >
        <div
          class="my-3 rounded product-card card"
        >
          <a
            href="/products/612bd7e14a0598527a005a69"
          >
            <img
              class="card-img-top p-1 fade-in"
              src="https://res.cloudinary.com/edeneault-cloudinary/image/upload/v1630923314/eShop/kjgv9cx9obl6gbvddohz.jpg"
            />
          </a>
          <div
            class="card-body"
          >
            <div
              class="list-group-flush py-2 list-group"
            >
              <div
                class="list-group-item"
              >
                <a
                  href="/products/612bd7e14a0598527a005a69"
                >
                  <div
                    class="d-block card-title"
                  >
                    <strong>
                       Blue Moon - Espresso Beans
                    </strong>
                  </div>
                </a>
              </div>
              <div
                class="list-group-item"
              >
                <div
                  class="card-text"
                >
                  <div
                    class="rating"
                  >
                    <span>
                      <i
                        class="fas fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                      <i
                        class="fas fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                      <i
                        class="fas fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                      <i
                        class="fas fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                      <i
                        class="fas fa-star-half-alt"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                       2 reviews
                    </span>
                  </div>
                </div>
              </div>
              <div
                class="price mx-auto list-group-item"
              >
                <h3
                  class="card-text"
                >
                  $8.99
                </h3>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col-xl-3 col-lg-4 col-md-6 col-sm-12"
      >
        <div
          class="my-3 rounded product-card card"
        >
          <a
            href="/products/6125d04db949259bdec4de39"
          >
            <img
              class="card-img-top p-1 fade-in"
              src="https://res.cloudinary.com/edeneault-cloudinary/image/upload/v1630783859/eShop/jqfbb27ruehuw9cpvwsp.jpg"
            />
          </a>
          <div
            class="card-body"
          >
            <div
              class="list-group-flush py-2 list-group"
            >
              <div
                class="list-group-item"
              >
                <a
                  href="/products/6125d04db949259bdec4de39"
                >
                  <div
                    class="d-block card-title"
                  >
                    <strong>
                       Lavazza - Super Crema Espresso
                    </strong>
                  </div>
                </a>
              </div>
              <div
                class="list-group-item"
              >
                <div
                  class="card-text"
                >
                  <div
                    class="rating"
                  >
                    <span>
                      <i
                        class="fas fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                      <i
                        class="fas fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                      <i
                        class="fas fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                      <i
                        class="fas fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                      <i
                        class="fas fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                       1 reviews
                    </span>
                  </div>
                </div>
              </div>
              <div
                class="price mx-auto list-group-item"
              >
                <h3
                  class="card-text"
                >
                  $39.99
                </h3>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col-xl-3 col-lg-4 col-md-6 col-sm-12"
      >
        <div
          class="my-3 rounded product-card card"
        >
          <a
            href="/products/6125d04db949259bdec4de38"
          >
            <img
              class="card-img-top p-1 fade-in"
              src="https://res.cloudinary.com/edeneault-cloudinary/image/upload/v1630784005/eShop/n7vc4qbhavt7gsjadkm7.jpg"
            />
          </a>
          <div
            class="card-body"
          >
            <div
              class="list-group-flush py-2 list-group"
            >
              <div
                class="list-group-item"
              >
                <a
                  href="/products/6125d04db949259bdec4de38"
                >
                  <div
                    class="d-block card-title"
                  >
                    <strong>
                       Damesi - Organic Espresso Italiano
                    </strong>
                  </div>
                </a>
              </div>
              <div
                class="list-group-item"
              >
                <div
                  class="card-text"
                >
                  <div
                    class="rating"
                  >
                    <span>
                      <i
                        class="fas fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                      <i
                        class="fas fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                      <i
                        class="fas fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                      <i
                        class="fas fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                      <i
                        class="far fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                       2 reviews
                    </span>
                  </div>
                </div>
              </div>
              <div
                class="price mx-auto list-group-item"
              >
                <h3
                  class="card-text"
                >
                  $11.99
                </h3>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col-xl-3 col-lg-4 col-md-6 col-sm-12"
      >
        <div
          class="my-3 rounded product-card card"
        >
          <a
            href="/products/6125d04db949259bdec4de3a"
          >
            <img
              class="card-img-top p-1 fade-in"
              src="https://res.cloudinary.com/edeneault-cloudinary/image/upload/v1630782951/eShop/whsbs3qytgxaw2sk8alm.jpg"
            />
          </a>
          <div
            class="card-body"
          >
            <div
              class="list-group-flush py-2 list-group"
            >
              <div
                class="list-group-item"
              >
                <a
                  href="/products/6125d04db949259bdec4de3a"
                >
                  <div
                    class="d-block card-title"
                  >
                    <strong>
                       Illy - Espresso 100% Arabica
                    </strong>
                  </div>
                </a>
              </div>
              <div
                class="list-group-item"
              >
                <div
                  class="card-text"
                >
                  <div
                    class="rating"
                  >
                    <span>
                      <i
                        class="fas fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                      <i
                        class="fas fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                      <i
                        class="fas fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                      <i
                        class="fas fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                      <i
                        class="fas fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                       1 reviews
                    </span>
                  </div>
                </div>
              </div>
              <div
                class="price mx-auto list-group-item"
              >
                <h3
                  class="card-text"
                >
                  $14.99
                </h3>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col-xl-3 col-lg-4 col-md-6 col-sm-12"
      >
        <div
          class="my-3 rounded product-card card"
        >
          <a
            href="/products/6125d04db949259bdec4de3c"
          >
            <img
              class="card-img-top p-1 fade-in"
              src="https://res.cloudinary.com/edeneault-cloudinary/image/upload/v1630783263/eShop/sqiwu5ummm4y1to3cp5o.jpg"
            />
          </a>
          <div
            class="card-body"
          >
            <div
              class="list-group-flush py-2 list-group"
            >
              <div
                class="list-group-item"
              >
                <a
                  href="/products/6125d04db949259bdec4de3c"
                >
                  <div
                    class="d-block card-title"
                  >
                    <strong>
                       Lavazza - Espresso - Medium Roast
                    </strong>
                  </div>
                </a>
              </div>
              <div
                class="list-group-item"
              >
                <div
                  class="card-text"
                >
                  <div
                    class="rating"
                  >
                    <span>
                      <i
                        class="fas fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                      <i
                        class="fas fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                      <i
                        class="fas fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                      <i
                        class="fas fa-star-half-alt"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                      <i
                        class="far fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                       10 reviews
                    </span>
                  </div>
                </div>
              </div>
              <div
                class="price mx-auto list-group-item"
              >
                <h3
                  class="card-text"
                >
                  $9.99
                </h3>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col-xl-3 col-lg-4 col-md-6 col-sm-12"
      >
        <div
          class="my-3 rounded product-card card"
        >
          <a
            href="/products/61306f847ef9b1460b0f6c75"
          >
            <img
              class="card-img-top p-1 fade-in"
              src="https://res.cloudinary.com/edeneault-cloudinary/image/upload/v1630785512/eShop/gvqcgmyd7wqdxexqte1t.jpg"
            />
          </a>
          <div
            class="card-body"
          >
            <div
              class="list-group-flush py-2 list-group"
            >
              <div
                class="list-group-item"
              >
                <a
                  href="/products/61306f847ef9b1460b0f6c75"
                >
                  <div
                    class="d-block card-title"
                  >
                    <strong>
                       Italian Coffee - Compatible pods
                    </strong>
                  </div>
                </a>
              </div>
              <div
                class="list-group-item"
              >
                <div
                  class="card-text"
                >
                  <div
                    class="rating"
                  >
                    <span>
                      <i
                        class="far fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                      <i
                        class="far fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                      <i
                        class="far fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                      <i
                        class="far fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                      <i
                        class="far fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                       0 reviews
                    </span>
                  </div>
                </div>
              </div>
              <div
                class="price mx-auto list-group-item"
              >
                <h3
                  class="card-text"
                >
                  $25
                </h3>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col-xl-3 col-lg-4 col-md-6 col-sm-12"
      >
        <div
          class="my-3 rounded product-card card"
        >
          <a
            href="/products/61306f7d7ef9b1460b0f6c6c"
          >
            <img
              class="card-img-top p-1 fade-in"
              src="https://res.cloudinary.com/edeneault-cloudinary/image/upload/v1630786349/eShop/yt55uakwnkvkfrxck7oj.jpg"
            />
          </a>
          <div
            class="card-body"
          >
            <div
              class="list-group-flush py-2 list-group"
            >
              <div
                class="list-group-item"
              >
                <a
                  href="/products/61306f7d7ef9b1460b0f6c6c"
                >
                  <div
                    class="d-block card-title"
                  >
                    <strong>
                       Espresso Capsules Sampler Bundle
                    </strong>
                  </div>
                </a>
              </div>
              <div
                class="list-group-item"
              >
                <div
                  class="card-text"
                >
                  <div
                    class="rating"
                  >
                    <span>
                      <i
                        class="fas fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                      <i
                        class="fas fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                      <i
                        class="fas fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                      <i
                        class="fas fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                      <i
                        class="fas fa-star"
                        style="color: rgb(140, 61, 61);"
                      />
                    </span>
                    <span>
                       1 reviews
                    </span>
                  </div>
                </div>
              </div>
              <div
                class="price mx-auto list-group-item"
              >
                <h3
                  class="card-text"
                >
                  $35.99
                </h3>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
